test/python/pulse/test_frames.py

```diff
@@ -0,0 +1,156 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Test cases for frames in Qiskit pulse."""
+
+import numpy as np
+
+from qiskit.circuit import Parameter
+from qiskit.compiler.assembler import frames_configuration
+import qiskit.pulse as pulse
+from qiskit.test import QiskitTestCase
+from qiskit.pulse.transforms import resolve_frames
+from qiskit.pulse.resolved_frame import ResolvedFrame
+from qiskit.pulse.parameter_manager import ParameterManager
+
+
+class TestFrame(QiskitTestCase):
+    """Basic frame tests."""
+
+    def basic(self):
+        """Test basic functionality of frames."""
+
+        frame = pulse.Frame(123)
+        self.assertEqual(frame.index, 123)
+        self.assertEqual(frame.name, 'f123')
+
+    def test_parameter(self):
+        """Test parameter assignment."""
+        parameter_manager = ParameterManager()
+        param = Parameter('a')
+        frame = pulse.Frame(param)
+        self.assertTrue(frame.is_parameterized())
+
+        parameter_manager.update_parameter_table(frame)
+        new_frame = parameter_manager.assign_parameters(frame, {param: 123})
+
+        self.assertEqual(new_frame, pulse.Frame(123))
+        self.assertEqual(frame, pulse.Frame(param))
+
+
+class TestResolvedFrames(QiskitTestCase):
+    """Test that resolved frames behave properly."""
+
+    def setUp(self):
+        super().setUp()
+        self.dt_ = 2.2222222222222221e-10
+        self.freq0 = 0.093e9  # Frequency of frame 0
+        self.freq1 = 0.125e9  # Frequency of frame 1
+
+    def test_phase_advance(self):
+        """Test that phases are properly set when frames are resolved."""
+
+        d0 = pulse.DriveChannel(0)
+        sig0 = pulse.Signal(pulse.Gaussian(160, 0.1, 40), pulse.Frame(0))
+        sig1 = pulse.Signal(pulse.Gaussian(160, 0.1, 40), pulse.Frame(1))
+
+        with pulse.build() as sched:
+            pulse.play(sig0, d0)
```

Comment:
Yes, the frames configuration allows one to determine which channels a frame broadcasts shift/set phase/frequency instructions to.

Reply:
I should have been more explicit. Let's not add any restrictions to the previously allowed instructions. Here's my question:

`Frame(0)` is configured to broadcast on `[DriveChannel(0), ControlChannel(3)]`
Is it wrong to say `Play(signal(waveform, Frame(0)), DriveChannel(3))`? This instruction wouldn't affect any other channel, right?

Reply:
This is really good point. Technically `Frame` in `Signal` just determine the phase factor at time t, so you can generate waveform on the frame0, and this can be played on any other channels. This makes me feel current signal formalism doesn't match with channel-based schedule timeslots formalism.

Reply:
I don't think a single frame can be broadcasted to multiple channels. This is currently supported because of control channels, but the control channel for cross resonance is implementation of frame (logical channel) on backend.

I think frame to channel mapping should be unique (but not for opposite, for example, qutrit), namely, channel and frame are 1:N mapping. However this can be N:M in current implementation, and such mapping is usually hard to manage. If the mapping is 1:N, we don't need to attach channel information to frame. 

This will drastically simplify the resolving logic:
1. Check play instruction. If the operand is signal, check associated frame.
2. Check frame operation allocated before the play instruction (preferably this is schedule block. with schedule we suffer the absolute time of those isntructions).
3.  Calculate phase accumulation wrt the frequency offset from channel frequency.
4. Insert shift phase and shift frequency before play instruction (frame is resolved here).
5. Apply transformation.

Reply:
Perhaps the best way of doing this is saying that we do not broadcast frame instructions to channels. So we need the following (which we already have):
* channel trackers which track the frequency and phase of physical channels.
* frame trackers which track the frequency and phase of frames.

Then when
```python
Play(Signal(waveform, Frame(X)), DriveChannel(Y))
```
is encountered we insert the appropriate phase shifts and frequency sets on DriveChannel(Y) by comparing the phase and frequency of the channel tracker to the frame tracker.
If a user starts mixing frame-based instructions with frameless instructions, i.e.
```python
Play(Signal(waveform1, Frame(X)), DriveChannel(Y))
Play(waveform2, DriveChannel(Y))
```
then we assume that the second waveform is played in `Frame(X)`. So the frame resolved schedule would look something like:
```python
t1, SetFrequency(freq_frame_x at t1, DriveChannel(Y))
t1, ShiftPhase(phase_frame_x - phase_drive_y at t1, DriveChannel(Y))
t1, Play(waveform1, DriveChannel(Y))
t2, Play(waveform2, DriveChannel(Y))
```
with `t2 = t1 + waveform1.duration`.

Reply:
> we insert the appropriate phase shifts and frequency sets on DriveChannel(Y) by comparing the phase and frequency of the channel tracker to the frame tracker.

This means we can drive measurement channels with a frame assigned to a drive channel. This causes an error due to available lo range. I still think unique frame-channel mapping is required.